File: api-java/src/main/java/fr/huiitre/tools/modules/riot/sync/application/usecase/SyncValorantContentTiersUseCase.java
package fr.huiitre.tools.modules.riot.sync.application.usecase;

import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import fr.huiitre.tools.modules.core.module.domain.ModuleCode;
import fr.huiitre.tools.modules.core.role.domain.RoleCode;
import fr.huiitre.tools.modules.core.security.application.usecase.SecuredUseCase;
import fr.huiitre.tools.modules.riot.sync.application.ValorantContentTierSyncData;
import fr.huiitre.tools.modules.riot.sync.application.ValorantSyncReport;
import fr.huiitre.tools.modules.riot.sync.application.ports.ValorantContentTierDataProvider;
import fr.huiitre.tools.modules.riot.sync.application.ports.ValorantContentTierSyncRepository;
import fr.huiitre.tools.modules.riot.valorant.application.catalog.view.ValorantContentTierView;

@Service
@Transactional
public class SyncValorantContentTiersUseCase implements SecuredUseCase {

    private final ValorantContentTierDataProvider contentTierDataProvider;
    private final ValorantContentTierSyncRepository contentTierSyncRepository;

    public SyncValorantContentTiersUseCase(
            ValorantContentTierDataProvider contentTierDataProvider,
            ValorantContentTierSyncRepository contentTierSyncRepository) {
        this.contentTierDataProvider = contentTierDataProvider;
        this.contentTierSyncRepository = contentTierSyncRepository;
    }

    @Override
    public Optional<ModuleCode> requiredModule() {
        return Optional.of(ModuleCode.RIOT);
    }

    @Override
    public RoleCode requiredRole() {
        return RoleCode.TECH;
    }

    public ValorantSyncReport execute() {
        Map<UUID, ValorantContentTierView> currentByAssetId = contentTierSyncRepository.findAll().stream()
                .collect(Collectors.toMap(ValorantContentTierView::assetId, it -> it));

        List<ValorantContentTierSyncData> external = contentTierDataProvider.fetchAll();

        Set<UUID> externalAssetIds = external.stream()
                .map(ValorantContentTierSyncData::getAssetId)
                .collect(Collectors.toSet());

        int created = 0;
        int updated = 0;
        int deleted = 0;

        for (ValorantContentTierSyncData ext : external) {
            ValorantContentTierView existing = currentByAssetId.get(ext.getAssetId());

            if (existing == null) {
                contentTierSyncRepository.save(ext);
                created++;
                continue;
            }

            boolean changed = !Objects.equals(existing.name(), ext.getName())
                    || !Objects.equals(existing.devName(), ext.getDevName())
                    || existing.rank() != ext.getRank()
                    || existing.juiceValue() != ext.getJuiceValue()
                    || existing.juiceCost() != ext.getJuiceCost()
                    || !Objects.equals(existing.highlightColor(), ext.getHighlightColor())
                    || !Objects.equals(existing.displayIconUrl(), ext.getDisplayIconUrl());

            if (changed) {
                contentTierSyncRepository.update(existing.id(), ext);
                updated++;
            }
        }

        for (ValorantContentTierView current : currentByAssetId.values()) {
            if (!externalAssetIds.contains(current.assetId())) {
                contentTierSyncRepository.delete(current.id());
                deleted++;
            }
        }

        return new ValorantSyncReport(created, updated, deleted);
    }
}

File: api/Modules/Riot/Sync/Application/Usecases/SyncValorantContentTiersUseCase.cs
using Tools.Api.Modules.Core.Security.Application.Services;
using Tools.Api.Modules.Core.Security.Application.Usecases;
using Tools.Api.Modules.Core.Security.Domain;
using Tools.Api.Modules.Riot.Sync.Application.Ports;

namespace Tools.Api.Modules.Riot.Sync.Application.Usecases;

public sealed class SyncValorantContentTiersUseCase(
    UseCaseAuthorizer authorizer,
    IValorantContentTierDataProvider contentTierDataProvider,
    IValorantContentTierSyncRepository contentTierSyncRepository
) : SecuredUseCase(authorizer)
{
    protected override RoleCode RequiredRole => RoleCode.Tech;
    protected override ModuleCode? RequiredModule => ModuleCode.Riot;

    public async Task<ValorantSyncReport> Execute()
    {
        var currentByAssetId = (await contentTierSyncRepository.FindAll())
            .ToDictionary(tier => tier.AssetId);

        var external = await contentTierDataProvider.FetchAll();
        var externalAssetIds = external.Select(tier => tier.AssetId).ToHashSet();

        var created = 0;
        var updated = 0;
        var deleted = 0;

        foreach (var tier in external)
        {
            if (!currentByAssetId.TryGetValue(tier.AssetId, out var existing))
            {
                await contentTierSyncRepository.Save(tier);
                created++;
                continue;
            }

            var changed = existing.Name != tier.Name
                || existing.DevName != tier.DevName
                || existing.Rank != tier.Rank
                || existing.JuiceValue != tier.JuiceValue
                || existing.JuiceCost != tier.JuiceCost
                || existing.HighlightColor != tier.HighlightColor
                || existing.DisplayIconUrl != tier.DisplayIconUrl;

            if (changed)
            {
                await contentTierSyncRepository.Update(existing.Id, tier);
                updated++;
            }
        }

        foreach (var current in currentByAssetId.Values.Where(tier => !externalAssetIds.Contains(tier.AssetId)))
        {
            await contentTierSyncRepository.Delete(current.Id);
            deleted++;
        }

        return new ValorantSyncReport(created, updated, deleted);
    }
}
